Adapted for biicode (C/C++ dependency manager)

franramirez688 opened the pull request:
Hi,

My name is Francisco Ramirez and I write you from **biicode** a small start-up that it’s developing a C and C++ dependency manager; just like Maven and Maven Central for Java. Our tool aims to allow our users to set-up a project with just #includes.

We're uploading as many interesting libraries as we can. I have forked your repository to support biicode. Our CMakeLists.txt code is on ``IF(BIICODE)`` statement, so it's not interfering with the original one (if you want to compile it without biicode).

I have tested it on Windows with MinGW and Visual Studio, Linux with GCC and OSX with Clang.

If you want to try it with biicode follow this steps:

1. [Install biicode](https://www.biicode.com/downloads)
2. Create new project folder:

    $ bii init flatbuffers
    $ cd flatbuffers

3. Checkout the github repository in block folder:

    $ git clone git@github.com:franramirez688/flatbuffers.git  blocks/google/flatbuffers
    $ bii cpp:build
    $ ./bin/google_flatbuffers_tests_test

You just created a biicode block named google/flatbuffers (username/blockname), you've compiled it and run an example. Now, if you'd want to publish it, [sign up](https://www.biicode.com/accounts/signup) and run the command:

    $ bii publish --tag STABLE 

By the momment, you could take a look to [this example](https://www.biicode.com/examples/flatbuffers), which is reusing your code (uploaded in [fenix account user](https://www.biicode.com/fenix/flatbuffers))

If you want to know more about biicode or need help just write me! francisco.ramirez@biicode.com

[Flatbuffers doc example](http://docs.biicode.com/c++/examples/flatbuffers.html)

Thanks you so much in advance

lilinghui commented:
great job

from my iPhone

> 在 2015年1月22日，23:41，Francisco Ramírez <notifications@github.com> 写道：
> 
> Hi,
> 
> My name is Francisco Ramirez and I write you from biicode a small start-up that it’s developing a C and C++ dependency manager; just like Maven and Maven Central for Java. Our tool aims to allow our users to set-up a project with just #includes.
> 
> We're uploading as many interesting libraries as we can. I have forked your repository to support biicode. Our CMakeLists.txt code is on IF(BIICODE) statement, so it's not interfering with the original one (if you want to compile it without biicode).
> 
> I have tested it on Windows with MinGW and Visual Studio, Linux with GCC and OSX with Clang.
> 
> If you want to try it with biicode follow this steps:
> 
> Install biicode
> Create new project folder:
> 
> $ bii init flatbuffers
> $ cd flatbuffers
> 
> Checkout the github repository in block folder:
> 
> $ git clone git@github.com:franramirez688/flatbuffers.git blocks/google/flatbuffers
> $ bii cpp:build
> $ ./bin/google_flatbuffers_tests_test
> 
> You just created a biicode block named google/flatbuffers (username/blockname), you've compiled it and run an example. Now, if you'd want to publish it, sign up and run the command:
> 
> $ bii publish --tag STABLE 
> By the momment, you could take a look to this example, which is reusing your code (uploaded in fenix account user)
> 
> If you want to know more about biicode or need help just write me! francisco.ramirez@biicode.com
> 
> Flatbuffers doc example
> 
> Thanks you so much in advance
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/google/flatbuffers/pull/133
> 
> Commit Summary
> 
> added biicode.conf and adapted CMakeLists.txt for using flatbuffers with biicode
> File Changes
> 
> M CMakeLists.txt (29)
> A biicode.conf (15)
> Patch Links:
> 
> https://github.com/google/flatbuffers/pull/133.patch
> https://github.com/google/flatbuffers/pull/133.diff
> —
> Reply to this email directly or view it on GitHub.
>

gwvo commented:
I like the idea of biicode and wouldn't mind including support for it. My concerns:

FlatBuffers itself doesn't have any dependencies, so seems the utility of biicode here is minimal (especially since we have to maintain the regular cmake code anyway). Users of FlatBuffers could benefit from biicode, but the default use of FlatBuffers involves just a single .h, so the only thing biicode would need to set up is its include dir. Then there's a more advanced use that involves several .cpp files (which shouldn't however be linked by default). The test code or samples should never be compiled when included in another project. Then of course the whole thing depends on a command line utility being compiled, to generate additional headers. I don't think your cmake code reflects the above. It is a very non-standard "library".

Second, assuming the above is addressed, the new cmake code appears to duplicate quite a bit of whats in the regular cmake code, which would worry me that we'd have to test with biicode and make changes in two places to ensure we're not breaking things for some users. The more the two parts can share things, the better.

franramirez688 commented:
Thanks a lot for your feedback! I'll update my MR with changes in the CMakeLists.txt to modify the minimum code, respecting the original one and adding a CI file (travis or appveyor) to check biicode building isn't breaking to make  the maintenance easier.

Give it a try and let us know what do you think

franramirez688 commented:
I modified CMakeLists.txt and biicode.conf to get a simpler solution. On the other hand, I've added a **.travis.yml** and **travis badge** to check building with biicode was right. The last ones are referenced to my biicode and github user, but if you decide to maintain your lib in biicode these would point to your account names

gwvo commented:
I'm not so sure about having the .travis.yml file, because people might expect it to build vanilla FlatBuffers, but now it builds it using biicode.

The cmake file looks better. Is all this copying necessary? I understand it supports building to a different directory, but the existing cmake file doesn't support that either.

stewartmiles commented on .travis.yml:
```diff
@@ -0,0 +1,36 @@
+
+install:
+  - sudo apt-get install libglu1-mesa-dev xorg-dev
+  - wget http://www.biicode.com/downloads/latest/ubuntu64
```
Why does this need to depend upon biicode, what are we getting for this?

stewartmiles commented on CMakeLists.txt:
```diff
@@ -114,3 +115,25 @@ if(FLATBUFFERS_BUILD_TESTS)
+endif(NOT BIICODE)
+
+IF(BIICODE)
+  INIT_BIICODE_BLOCK()
```
I'm not a big fan of distribution specific code in the main cmakelists file.  If this was factored out so that biicode was part of an optionally included file it may not be so bad.  Right now if we start adding rules for each distribution mechanism for the library this could get out of hand pretty quickly.

stewartmiles commented on biicode.conf:
```diff
@@ -0,0 +1,4 @@
+[parent]
+
+[paths]
+	include
```
What is this file for?  Could you add comments?

Would it be possible to put everything related to biicode in a subdirectory for example?

stewartmiles commented on readme.md:
```diff
@@ -1,4 +1,4 @@
-FlatBuffers Version 1.0
+FlatBuffers Version 1.0 [![Build Status](https://travis-ci.org/franramirez688/flatbuffers.svg?branch=master)](https://travis-ci.org/franramirez688/flatbuffers)
```
If we're going to setup Travis for Flatbuffers we should probably have it associated with a Flatbuffers maintainer's account instead.

stewartmiles commented:
Wouter asked me to take a look at this and I'm a little concerned how this could become messy pretty quickly.  In terms of using biicode in the Travis project, I guess that's ok as long as it's factored out into a separate section of the travis.yml at the moment the primary install target installs biicode which seems really odd.  As I mention in my comments on the CMakeLists.txt file, it would be more scalable - AFAIK - to put the biicode stuff in a subdirectory so that when someone comes along with support for the next package management hotness we don't just lump it into the core project file.

Does this make sense?  Sorry to churn you on this, we would just like to get this right.

Cheers,
Stewart

franramirez688 commented:
I just rebased and squashed my commits. I modified the original CMakeLists.txt and added a biicode.cmake into *CMake/* folder to separate biicode's building. I commented the biicode.conf file too.
.travis.yml file to build with biicode has been currently removed but I can replace it for a .travis.yml file to build "as usual"

Hope this fits you best,
Cheers,
Fran

stewartmiles commented:
Looking way better, any chance you could squash this series into one commit (git rebase -i) ?

What happened to the Travis build config?  Going to keep that in a fork?

franramirez688 commented:
I removed Travis file because I'm working on a flatbuffers' build config with and without biicode, so, you could check both at same building time. As soon as it's ready, I'll open another PR or add it here to merge it.

Cheers,
Fran

gwvo commented:
cherry-picked in https://github.com/google/flatbuffers/commit/da0f096ba220ec49bafc0743773a46524dd79832
